doc/user/content/transform-data/optimization.md

```diff
@@ -294,6 +294,87 @@ The following are the possible index usage types:
 - `fast path limit`: When a [fast path](/sql/explain/#fast-path-queries) query has a `LIMIT` clause but no `ORDER BY` clause, then Materialize will read from the index only as many records as required to satisfy the `LIMIT` (plus `OFFSET`) clause.
 
+## Query Hints
+
+Materialize has at present one important [query hint]: the `EXPECTED GROUP SIZE`. This hint applies to indexed or materialized views that need to incrementally maintain [`MIN`], [`MAX`], or [Top K] query patterns. Maintaining these query patterns while delivering low latency under updates is demanding in terms of main memory. This is because Materialize builds a hierarchy of [arrangements] so that data can be partitioned and each update can trickle through only a small number of levels in the hierarchy.
+
+The number of levels needed in the hierarchical scheme is by default set assuming uneven partitioning wherein a large number of updates could hit a single query group, as specified, e.g., by `GROUP BY`, `LATERAL`, or `DISTINCT ON`. By specifying the `EXPECTED GROUP SIZE`, it is possible to refine this assumption, allowing Materialize to build a hierarchy with less levels and lower memory consumption without sacrificing update latency.
+
+Consider the previous example with the collection `sections`. Maintenance of the maximum `course_id` per `teacher` can be achieved with a materialized view:
+
+```sql
+CREATE MATERIALIZED VIEW max_course_id_per_teacher AS
+SELECT teacher_id, MAX(course_id)
+FROM sections
+GROUP BY teacher_id;
+```
+
+If the largest number of `course_id` values that are allocated to a single `teacher_id` is known, then this number can be provided as the `EXPECTED GROUP SIZE`. For the query above, it is possible to get an estimate for this number by:
+
+```sql
+SELECT MAX(course_count)
+FROM (
+  SELECT teacher_id, COUNT(*) course_count
+  FROM sections
+  GROUP BY teacher_id
+);
+```
+
+However, the estimate is based only on data that is already present in the system. So taking into account how much this largest number could expand is critical to avoid issues with update latency after tuning the `EXPECTED GROUP SIZE`.
+
+For our example, let's suppose that we determined the largest number of courses per teacher to be `1000`. Then, the original definition of `max_course_id_per_teacher` can be revised to include the `EXPECTED GROUP SIZE` query hint as follows:
+
+```sql
+CREATE MATERIALIZED VIEW max_course_id_per_teacher AS
+SELECT teacher_id, MAX(course_id)
+FROM sections
+GROUP BY teacher_id
+OPTIONS (EXPECTED GROUP SIZE = 1000)
+```
+
+The same hint can be provided in [Top K] query patterns specified by `DISTINCT ON` or `LATERAL`. As examples, consider that we wish not to compute the maximum `course_id`, but rather the `id` of the section of this top course. This computation can be incrementally maintained by the following materialized view:
```

Comment:
> Top K

In `mz_internal.md` it's written as "top-k". It would be good to unify.

Reply:
Cool, changed to Top K in `mz_internal.md`.